File: frontend-2024-grupo2/src/app/data/models/Ticket.ts
import { SeatType, SeatTypeJSON } from "./SeatType"

export type TicketJSON = {
	seatType: SeatTypeJSON
	price: number
}

export class Ticket {
  
	constructor(
		public seatType: SeatType,
		public price: number
	) {}

	static fromJSON(ticketJSON: TicketJSON): Ticket {
		return Object.assign(new Ticket(
			SeatType.fromJSON(ticketJSON.seatType),
			ticketJSON.price
		))
	}

	toJSON(): TicketJSON {
		return {
			seatType: this.seatType.toJSON(),
			price: this.price
		}
	}
}

File: backend-2024-grupo2/src/main/kotlin/ar/edu/unsam/phm/backendgrupo2/domain/Ticket.kt
package ar.edu.unsam.phm.backendgrupo2.domain

import ar.edu.unsam.phm.backendgrupo2.dto.TicketDTO
import jakarta.persistence.*
import java.math.BigDecimal

@Entity
@Table(name = "ticket")
class Ticket {

  companion object {
    fun fromJSON(ticketDTO: TicketDTO, functionShow: FunctionShow): Ticket =
      Ticket().apply {
        this.id = ticketDTO.id
        this.seatType = SeatType.fromJSON(ticketDTO.seatType)
        this.functionShow = functionShow
      }
  }

  @Id
  @GeneratedValue
  var id: Int = 0

  @ManyToOne(fetch = FetchType.EAGER)
  var seatType: SeatType? = null

  @ManyToOne(fetch = FetchType.EAGER)
  var functionShow: FunctionShow? = null
  
  fun cost(): BigDecimal {
    val placeOfShow: PlaceOfShow = this.functionShow?.show?.placeOfShow!!
    return if (placeOfShow.amountOfPlaces() != 0)
      (placeOfShow.locationCost(this) + this.functionShow?.show?.costBand!!) /
        (placeOfShow.amountOfPlaces().toBigDecimal())
    else 0.0.toBigDecimal()
  }

  fun price(): BigDecimal =
    this.cost() * this.functionShow?.show!!.status?.profit()!!
  
  fun hasComment(comment: Comment): Boolean =
    this.functionShow?.show?.hasComment(comment)!!
}